src/conventions/h-group/take-action.js

```diff
@@ -230,7 +269,57 @@ export function take_action(game) {
 	let best_play_clue, clue_value;
 	if (state.clue_tokens > 0) {
 		const all_play_clues = play_clues.flat();
-		({ clue: best_play_clue, clue_value } = select_play_clue(all_play_clues));
+		let consider_clues = all_play_clues;
+		const chop = game.me.chop(state.hands[state.ourPlayerIndex]);
+		let saved_clue;
+		let saved_clue_value = -99;
+
+		// Consider saving clues to finesse positions for players who likely have
+		// better cards on chop.
+		if (!state.inEndgame() && !state.early_game && state.clue_tokens < 4 && chop !== undefined) {
+			const our_chop_value = cardValue(state, game.me, chop, chop.order);
+			const better_givers = [];
+			for (let i = 0; i < state.numPlayers; ++i) {
+				if (i == state.ourPlayerIndex)
+					continue;
+				const player = game.players[i];
+				const hand = state.hands[i];
+				if (player.thinksLoaded(state, i, {assume: false}))
+					continue;
+				const otherChop = player.chop(hand);
+				// Saves clue for locked players or players who likely have a better chop.
+				if (otherChop === undefined && player.thinksLocked(state, i))
```

Comment:
Are both of these conditions necessary? It seems like not having a chop implies they are locked, since we already eliminated the loaded possibility. That would simplify the else condition too.

In general, I also prefer a more functional approach, using `Array.filter()` or `Array.map()` over `for` loops when there are no side effects. I find it more concise and easier to understand when things are defined as the flow of data rather than a series of instructions. For example, `better_givers` can be written in this form:
```js
const better_givers = Utils.range(0, state.numPlayers).filter(i => {
    const player = game.players[i];
    const otherChop = player.chop(state.hands[i]);

    return i !== state.ourPlayerIndex && !player.thinksLoaded(state, i, {assume: false}) &&
        (chop === undefined || cardValue(state, player, otherChop, otherChop.order) >= our_chop_value);
});
```
which "defines" the array rather than building it with mutation. This also lets TypeScript infer the type of `better_givers` without us needing to explicitly state it.

Reply:
> Are both of these conditions necessary? It seems like not having a chop implies they are locked, since we already eliminated the loaded possibility. That would simplify the else condition too.

I may be mistaken but It looks to me that chopIndex doesn't take into account know trash, in which case the player would be neither locked nor have a chop.

> In general, I also prefer a more functional approach, using `Array.filter()` or `Array.map()` over `for` loops when there are no side effects. I find it more concise and easier to understand when things are defined as the flow of data rather than a series of instructions. For example, `better_givers` can be written in this form:
> 
> ```js
> const better_givers = Utils.range(0, state.numPlayers).filter(i => {
>     const player = game.players[i];
>     const otherChop = player.chop(state.hands[i]);
> 
>     return i !== state.ourPlayerIndex && !player.thinksLoaded(state, i, {assume: false}) &&
>         (otherChop === undefined || cardValue(state, player, otherChop, otherChop.order) >= our_chop_value);
> });
> ```
> 
> which "defines" the array rather than building it with mutation. This also lets TypeScript infer the type of `better_givers` without us needing to explicitly state it.

That's fair, I'm still getting used to using functional construction more often :-). Done.

Reply:
> I may be mistaken but It looks to me that chopIndex doesn't take into account know trash, in which case the player would be neither locked nor have a chop.

If a player has known trash then they'd be loaded (having either a known playable or known trash) and would be caught by the earlier condition.

Reply:
Ah, right! That makes sense! Switched it to your suggested code which I believe captures all of the necessary conditions.